src/monitor.py

```diff
@@ -115,3 +119,147 @@ def read_recent_logs(self, job: JobDto, tail: int = 20) -> str:
 
     def _get_internal_job_url(self, job: JobDto) -> str:
         return f'http://{job.internal_name}'
+
+
+    def check_deployment_for_early_errors(self, resource_name: str):
+        """
+        Check for indicators of early errors in the deployment
+        If errors were found, raises an error
+        If no errors were found, returns
+        If unsure, continues checking until timeout, then returns
+        """
+        # There is not a single way to determine if a deployment has failed early, so we combine multiple types of checks,
+        # and require that consecutive checks agree the deployment has failed or passed.
+        # A "check" return True on deployment success, False on deployment failure, and None if unsure.
+        # The timeouts, number of checks, and loop_sleeps are mostly based on heuristics and vibes:
+        #   We don't want to spam the cluster with requests
+        #   But we also do not want the deployment to take much longer
+        # If the deployment just works, it will be delayed by O(n_consecutive_checks * total_loop_time)
+        # TODO: If we are really confident in 'kubectl rollout status' we could spawn it in a separate thread and wait for it to finish with a watch='true'
+        n_consecutive_checks = 3
```

Comment:
In my opinion, there's no point to keep it other than 1, but I could be wrong, so I'm trying to understand this. If there's insufficient memory in a cluster, it won't fix itself in a few seconds. Is there a proven reason for this, or is it just thinking ahead? Without it, it would simplify the logic and, most importantly, speed up deployment (not having to wait unnecessarily)

Reply:
It's there because I was worried about transient false positives, i.e. a "failing"-indicating event that was recoverable. It stems mainly from me not knowing Kubernetes super well.

Reply:
But otherwise, I agree. I'll change it to a single check.